builtin/credential/approle/secret_id.go

```diff
@@ -0,0 +1,500 @@
+	"sync"
+	"time"
+
+	"github.com/hashicorp/go-uuid"
+	"github.com/hashicorp/vault/helper/strutil"
+	"github.com/hashicorp/vault/logical"
+	"github.com/hashicorp/vault/logical/framework"
+)
+
+// secretIDStorageEntry represents the information stored in storage
+// when a SecretID is created. The structure of the SecretID storage
+// entry is the same for all the types of SecretIDs generated.
+type secretIDStorageEntry struct {
+	// Accessor for the SecretID. It is a random UUID serving as
+	// a secondary index for the SecretID. This uniquely identifies
+	// the SecretID it belongs to, and hence can be used for listing
+	// and deleting SecretIDs. Accessors cannot be used as valid
+	// SecretIDs during login.
+	SecretIDAccessor string `json:"secret_id_accessor" structs:"secret_id_accessor" mapstructure:"secret_id_accessor"`
+
+	// Number of times this SecretID can be used to perform the login operation
+	SecretIDNumUses int `json:"secret_id_num_uses" structs:"secret_id_num_uses" mapstructure:"secret_id_num_uses"`
+
+	// Duration after which this SecretID should expire. This is
+	// croleed by the backend mount's max TTL value.
+	SecretIDTTL time.Duration `json:"secret_id_ttl" structs:"secret_id_ttl" mapstructure:"secret_id_ttl"`
+
+	// The time in UTC when the SecretID was created
+	CreationTime time.Time `json:"creation_time" structs:"creation_time" mapstructure:"creation_time"`
+
+	// The time in UTC when the SecretID becomes eligible for tidy
+	// operation. Tidying is performed by the PeriodicFunc of the
+	// backend which is 1 minute apart.
+	ExpirationTime time.Time `json:"expiration_time" structs:"expiration_time" mapstructure:"expiration_time"`
+
+	// The time in UTC representing the last time this storage entry was modified
+	LastUpdatedTime time.Time `json:"last_updated_time" structs:"last_updated_time" mapstructure:"last_updated_time"`
+
+	// Metadata that belongs to the SecretID.
+	Metadata map[string]string `json:"metadata" structs:"metadata" mapstructure:"metadata"`
+}
+
+// Represents the payload of the storage entry of the accessor that maps to a unique
+// SecretID. Note that SecretIDs should never be stored in plaintext anywhere in the
+// backend. SecretIDHMAC will be used as an index to fetch the properties of the
+// SecretID and to delete the SecretID.
+type secretIDAccessorStorageEntry struct {
+	// Hash of the SecretID which can be used to find the storage index at which
+	// properties of SecretID is stored.
+	SecretIDHMAC string `json:"secret_id_hmac" structs:"secret_id_hmac" mapstructure:"secret_id_hmac"`
+}
+
+// roleIDStorageEntry represents the reverse mroleing from RoleID to Role
+type roleIDStorageEntry struct {
+	Type string `json:"type" structs:"type" mapstructure:"type"`
+	Name string `json:"name" structs:"name" mapstructure:"name"`
+}
+
+// setRoleIDEntry creates a storage entry that maps RoleID to Role
+func (b *backend) setRoleIDEntry(s logical.Storage, roleID string, roleIDEntry *roleIDStorageEntry) error {
```

Comment:
In the wake of mutable role IDs, probably need to make sure that this is doing the right thing. Or if it's even needed, if the path is going to use the hmac of the role name.